src/bp/core/services/dialog/engine.ts

```diff
@@ -180,126 +172,90 @@ export class DialogEngine {
       throw new Error(`Could not find any timeout node for session "${sessionId}"`)
     }
 
-    const context = {
-      currentNodeName: timeoutNode.name,
-      currentFlowName: timeoutFlow.name
-    }
+    event.state.context.currentNode = timeoutNode.name
+    event.state.context.currentFlow = timeoutFlow.name
 
-    await this._updateContext(sessionId, context)
-    await this.processEvent(sessionId, event)
+    return await this.processEvent(sessionId, event)
   }
 
-  private async _getOrCreateSession(sessionId, event) {
-    const session = await this.sessionService.getSession(sessionId)
+  private initializeContext(event) {
+    this._logStart(event.botId)
 
-    if (!session) {
-      this._logStart(event.botId)
+    const defaultFlow = this._findFlow(event.botId, 'main.flow.json')
+    const startNode = this._findNode(defaultFlow, defaultFlow.startNode)
 
-      // Create the new session context
-      const defaultFlow = this._findFlow(event.botId, 'main.flow.json')
-      const startNode = this._findNode(defaultFlow, defaultFlow.startNode)
-      const context = {
-        currentNodeName: startNode.name,
-        currentFlowName: defaultFlow.name
-      }
-      const emptyState = {}
-      return this.sessionService.createSession(sessionId, event.botId, emptyState, context, event)
+    event.state.context = {
```

Comment:
I don't understand why the context is in the state now. The state is exposed to the user. The context should not because it contains engine specific data that you dont want the user to mess with. 

We could "freeze" the context as we discussed.